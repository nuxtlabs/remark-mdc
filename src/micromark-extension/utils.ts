import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
