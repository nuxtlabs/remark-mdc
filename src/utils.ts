export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
