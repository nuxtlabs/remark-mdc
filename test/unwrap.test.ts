import { describe, expect } from 'vitest'
import { runMarkdownTests } from './utils'

describe('auto-unwrap', () => {
  runMarkdownTests({
    simple: {
      mdcOptions: {
        experimental: {
          autoUnwrap: true
        }
      },
      markdown: '::test\nHello World\n::',
      extra (_markdown, ast, _expected) {
        expect(ast.children[0].children.length).toBe(1)
        expect(ast.children[0].children[0]).property('type', 'text')
        expect(ast.children[0].children[0]).property('value', 'Hello World')
      }
    },
    slot: {
      mdcOptions: {
        experimental: {
          autoUnwrap: true
        }
      },
      markdown: '::test\n#title\nHello World\n::',
      extra (_markdown, ast, _expected) {
        const slot = ast.children[0].children[0]
        expect(slot.children.length).toBe(1)
        expect(slot.children[0]).property('type', 'text')
        expect(slot.children[0]).property('value', 'Hello World')
      }
    },
    twoSlots: {
      mdcOptions: {
        experimental: {
          autoUnwrap: true
        }
      },
      markdown: '::test\nFoo bar\n\n#title\nHello World\n::',
      extra (_markdown, ast, _expected) {
        const nonSlotChild = ast.children[0].children[0]
        expect(nonSlotChild).property('type', 'text')
        expect(nonSlotChild).property('value', 'Foo bar')

        const slot = ast.children[0].children[1]
        expect(slot.children.length).toBe(1)
        expect(slot.children[0]).property('type', 'text')
        expect(slot.children[0]).property('value', 'Hello World')
      }
    },
    twoChildren: {
      mdcOptions: {
        experimental: {
          autoUnwrap: true
        }
      },
      markdown: '::test\nFoo bar\n\nHello World\n::',
      extra (_markdown, ast, _expected) {
        for (const child of ast.children[0].children) {
          expect(child).property('type', 'paragraph')
        }
      }
    },
    componentChildren: {
      mdcOptions: {
        experimental: {
          autoUnwrap: true
        }
      },
      markdown: '::test\n  :::another-component\n  :::\n::',
      extra (_markdown, ast, _expected) {
        expect(ast.children[0].children.length).toBe(1)
        expect(ast.children[0].children[0]).property('type', 'containerComponent')
      }
    }
  })
})
