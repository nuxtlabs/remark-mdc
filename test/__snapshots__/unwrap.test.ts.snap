// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auto-unwrap > componentChildren 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {},
          "children": [],
          "data": {
            "hName": "another-component",
            "hProperties": {},
          },
          "fmAttributes": {},
          "name": "another-component",
          "position": {
            "end": {
              "column": 6,
              "line": 3,
              "offset": 35,
            },
            "start": {
              "column": 3,
              "line": 2,
              "offset": 9,
            },
          },
          "type": "containerComponent",
        },
      ],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "test",
      "position": {
        "end": {
          "column": 3,
          "line": 4,
          "offset": 38,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 4,
      "offset": 38,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`auto-unwrap > simple 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 2,
              "offset": 18,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 7,
            },
          },
          "type": "text",
          "value": "Hello World",
        },
      ],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "mdc": {
        "unwrapped": "paragraph",
      },
      "name": "test",
      "position": {
        "end": {
          "column": 3,
          "line": 3,
          "offset": 21,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 3,
      "offset": 21,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`auto-unwrap > slot 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "attributes": {},
          "children": [
            {
              "position": {
                "end": {
                  "column": 12,
                  "line": 3,
                  "offset": 25,
                },
                "start": {
                  "column": 1,
                  "line": 3,
                  "offset": 14,
                },
              },
              "type": "text",
              "value": "Hello World",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:title": "",
            },
          },
          "mdc": {
            "unwrapped": "paragraph",
          },
          "name": "title",
          "position": {
            "end": {
              "column": 3,
              "line": 4,
              "offset": 28,
            },
            "start": {
              "column": 2,
              "line": 2,
              "offset": 8,
            },
          },
          "type": "componentContainerSection",
        },
      ],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "test",
      "position": {
        "end": {
          "column": 3,
          "line": 4,
          "offset": 28,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 4,
      "offset": 28,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`auto-unwrap > twoChildren 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 8,
                  "line": 2,
                  "offset": 14,
                },
                "start": {
                  "column": 1,
                  "line": 2,
                  "offset": 7,
                },
              },
              "type": "text",
              "value": "Foo bar",
            },
          ],
          "position": {
            "end": {
              "column": 8,
              "line": 2,
              "offset": 14,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 7,
            },
          },
          "type": "paragraph",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 12,
                  "line": 4,
                  "offset": 27,
                },
                "start": {
                  "column": 1,
                  "line": 4,
                  "offset": 16,
                },
              },
              "type": "text",
              "value": "Hello World",
            },
          ],
          "position": {
            "end": {
              "column": 12,
              "line": 4,
              "offset": 27,
            },
            "start": {
              "column": 1,
              "line": 4,
              "offset": 16,
            },
          },
          "type": "paragraph",
        },
      ],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "name": "test",
      "position": {
        "end": {
          "column": 3,
          "line": 5,
          "offset": 30,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 5,
      "offset": 30,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`auto-unwrap > twoSlots 1`] = `
{
  "children": [
    {
      "attributes": {},
      "children": [
        {
          "position": {
            "end": {
              "column": 8,
              "line": 2,
              "offset": 14,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 7,
            },
          },
          "type": "text",
          "value": "Foo bar",
        },
        {
          "attributes": {},
          "children": [
            {
              "position": {
                "end": {
                  "column": 12,
                  "line": 5,
                  "offset": 34,
                },
                "start": {
                  "column": 1,
                  "line": 5,
                  "offset": 23,
                },
              },
              "type": "text",
              "value": "Hello World",
            },
          ],
          "data": {
            "hName": "component-slot",
            "hProperties": {
              "v-slot:title": "",
            },
          },
          "mdc": {
            "unwrapped": "paragraph",
          },
          "name": "title",
          "position": {
            "end": {
              "column": 3,
              "line": 6,
              "offset": 37,
            },
            "start": {
              "column": 2,
              "line": 4,
              "offset": 17,
            },
          },
          "type": "componentContainerSection",
        },
      ],
      "data": {
        "hName": "test",
        "hProperties": {},
      },
      "fmAttributes": {},
      "mdc": {
        "unwrapped": "paragraph",
      },
      "name": "test",
      "position": {
        "end": {
          "column": 3,
          "line": 6,
          "offset": 37,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "rawData": undefined,
      "type": "containerComponent",
    },
  ],
  "position": {
    "end": {
      "column": 3,
      "line": 6,
      "offset": 37,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
